Controller/Callback/Offsite.php

```diff
@@ -106,18 +106,163 @@ public function execute()
+                return $this->handleSuccess($order, $charge->id, $payment, $paymentMethod);
+            }
+
+            $this->handlePending($order, $payment);
+        } catch (Exception $e) {
+            $this->cancel($order, $e->getMessage());
+
+            return $this->redirect(self::PATH_CART);
+        }
+    }
+
+    /**
+     * Mark order as failed
+     *
+     * @param obejct $charge
+     * @param boolean $shopeePayFaild {TODO: Remove this once backend issue is fixed}
+     */
+    private function handleFailure($charge, $shopeePayFailed)
+    {
+        // restoring the cart
+        $this->checkoutSession->restoreQuote();
+        $failureMessage = $charge->failure_message ?
+            ucfirst($charge->failure_message) :
+            'Payment cancelled';
+        $errorMessage = __(
+            "Payment failed. $failureMessage, please contact our support if you have any questions."
+        );
+
+        // pass shopeePayFailed to avoid webhook to cancel payment
+        return $this->processFailedCharge($errorMessage, $shopeePayFailed);
+    }
+
+    /**
+     * Mark order as success
+     *
+     * @param object $order
+     * @param string $chargeId
+     * @param object $payment
+     * @param string $paymentMethod
+     */
+    private function handleSuccess($order, $chargeId, $payment, $paymentMethod)
+    {
+        // Update order state and status.
+        $order->setState(Order::STATE_PROCESSING);
+        $order->setStatus($order->getConfig()->getStateDefaultStatus(Order::STATE_PROCESSING));
+
+        $invoice = $this->helper->createInvoiceAndMarkAsPaid($order, $chargeId);
+        $this->emailHelper->sendInvoiceAndConfirmationEmails($order);
+
+        $paymentMethodLabel = $this->helper->getOmiseLabelByOmiseCode($paymentMethod);
+        // Add transaction.
+        $payment->addTransactionCommentsToOrder(
+            $payment->addTransaction(Transaction::TYPE_PAYMENT, $invoice),
+            __(
+                "Amount of %1 has been paid via Opn Payments $paymentMethodLabel payment",
+                $order->getBaseCurrency()->formatTxt($invoice->getBaseGrandTotal())
+            )
+        );
+
+        $order->save();
+        return $this->redirect(self::PATH_SUCCESS);
+    }
+
+    /**
+     * Mark order a pending
+     *
+     * @param object $order
+     * @param object $payment
+     */
+    private function handlePending($order, $payment)
+    {
+        // Update order state and status.
+        $order->setState(Order::STATE_PAYMENT_REVIEW);
+        $order->setStatus($order->getConfig()->getStateDefaultStatus(Order::STATE_PAYMENT_REVIEW));
+
+        // Add transaction.
+        $transaction = $payment->addTransaction(Transaction::TYPE_PAYMENT);
+        $transaction->setIsClosed(false);
+        $payment->addTransactionCommentsToOrder(
+            $transaction,
+            __('The payment has been processing.<br/>Due to the Bank process, this might takes a few seconds
```

Comment:
The language here is not great. Can we have @muthuswamyopn look at it to improve the grammar?

Reply:
Sure. I have DMed @muthuswamyopn.